File: Week 02/id_153/LeetCode_242_153.js
const isAnagram1 = (s, t) => {
  let sArr = s.split("");
  let tArr = t.split("");
  sArr.sort();
  tArr.sort();
  return sArr + "" === tArr + "";
};

const isAnagram2 = (s, t) => {
  const map = new Map();
  for (let i = 0; i < s.length; i++) {
    if (map.has(s[i])) {
      map.set(s[i], map.get(s[i]) + 1);
    } else {
      map.set(s[i], 1);
    }
  }

  for (let i = 0; i < t.length; i++) {
    if (map.has(t[i])) {
      map.set(t[i], map.get(t[i]) - 1);
    } else {
      map.set(t[i], 1);
    }

    if (map.get(t[i]) === 0) {
      map.delete(t[i]);
    }
  }
  if (map.size !== 0) {
    return false;
  }
  return true;
};

File: Week 02/id_153/LeetCode_242_153.java
import java.util.Arrays;

/**
 * LeetCode_242_153 https://leetcode-cn.com/problems/valid-anagram/
 */
public class LeetCode_242_153 {

  /**
   * 如果两个字符串长度不相等，一定不是异或词。 将字符串排序，排序后的结果相等，则为异或词
   * 
   * @param s
   * @param t
   * @return
   */
  public boolean isAnagram1(String s, String t) {
    if (s.length() != t.length()) {
      return false;
    }
    char[] str1 = s.toCharArray();
    char[] str2 = t.toCharArray();
    Arrays.sort(str1);
    Arrays.sort(str2);
    return Arrays.equals(str1, str2);
  }

  public boolean isAnagram2(String s, String t) {
    if (s.length() != t.length()) {
      return false;
    }

    int[] alpha = new int[26];
    for (int i = 0; i < s.length(); i++) {
      alpha[s.charAt(i) - 'a']++;
      alpha[t.charAt(i) - 'a']--;
    }

    for (int i = 0; i < alpha.length; i++) {
      if (alpha[i] != 0) {
        return false;
      }
    }
    return true;
  }
}